From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rui Marques <rui.marques@appentra.com>
Date: Wed, 19 Mar 2025 16:08:27 +0100
Subject: Hide `parenthesized_expression`

This commit must be removed now that we have semantic access to parent
nodes!!
---
 grammar.js                  |  28 +-
 test/corpus/constructs.txt  |   7 +-
 test/corpus/expressions.txt | 108 ++++----
 test/corpus/regressions.txt |  41 ++-
 test/corpus/statements.txt  | 533 ++++++++++++++++++------------------
 5 files changed, 351 insertions(+), 366 deletions(-)

diff --git a/grammar.js b/grammar.js
index 8d859d4..4de79f8 100644
--- a/grammar.js
+++ b/grammar.js
@@ -84,7 +84,7 @@ module.exports = grammar({
 
   conflicts: $ => [
     [$._expression, $.complex_literal],
-    [$._argument_list, $.parenthesized_expression],
+    [$._argument_list, $._parenthesized_expression],
     [$.case_statement],
     [$.data_set, $._expression],
     [$.data_statement, $.identifier],
@@ -203,7 +203,7 @@ module.exports = grammar({
       $.preproc_defined,
       alias($.preproc_unary_expression, $.unary_expression),
       alias($.preproc_binary_expression, $.binary_expression),
-      alias($.preproc_parenthesized_expression, $.parenthesized_expression),
+      alias($.preproc_parenthesized_expression, $._parenthesized_expression),
     ),
 
     preproc_parenthesized_expression: $ => seq(
@@ -952,7 +952,7 @@ module.exports = grammar({
 
     kind: $ => choice(
       seq(optional(alias('*', $.assumed_size)), $._argument_list),
-      seq('*', choice(alias(/\d+/, $.number_literal), $.parenthesized_expression))
+      seq('*', choice(alias(/\d+/, $.number_literal), $._parenthesized_expression))
     ),
 
     character_length: $ => seq(
@@ -1252,7 +1252,7 @@ module.exports = grammar({
     )),
 
     while_statement: $ => seq(caseInsensitive('while'),
-      $.parenthesized_expression),
+      $._parenthesized_expression),
 
     concurrent_statement: $ => seq(
       $.concurrent_header,
@@ -1311,13 +1311,13 @@ module.exports = grammar({
 
     inline_if_statement: $ => prec.right(2, seq(
       caseInsensitive('if'),
-      $.parenthesized_expression,
+      $._parenthesized_expression,
       $._statements
     )),
 
     arithmetic_if_statement: $ => prec.right(seq(
       caseInsensitive('if'),
-      $.parenthesized_expression,
+      $._parenthesized_expression,
       $.statement_label_reference,
       ',',
       $.statement_label_reference,
@@ -1328,7 +1328,7 @@ module.exports = grammar({
     block_if_statement: $ => seq(
       optional($.block_label_start_expression),
       caseInsensitive('if'),
-      $.parenthesized_expression,
+      $._parenthesized_expression,
       caseInsensitive('then'),
       optional($._block_label),
       $.end_of_statement,
@@ -1346,7 +1346,7 @@ module.exports = grammar({
 
     elseif_clause: $ => seq(
       whiteSpacedKeyword('else', 'if'),
-      $.parenthesized_expression,
+      $._parenthesized_expression,
       caseInsensitive('then'),
       optional($._block_label),
       $.end_of_statement,
@@ -1367,14 +1367,14 @@ module.exports = grammar({
 
     inline_where_statement: $ => prec.right(seq(
       caseInsensitive('where'),
-      $.parenthesized_expression,
+      $._parenthesized_expression,
       $._statements
     )),
 
     block_where_statement: $ => seq(
       optional($.block_label_start_expression),
       caseInsensitive('where'),
-      $.parenthesized_expression,
+      $._parenthesized_expression,
       $.end_of_statement,
       repeat($._statement),
       repeat($.elsewhere_clause),
@@ -1388,7 +1388,7 @@ module.exports = grammar({
 
     elsewhere_clause: $ => seq(
       whiteSpacedKeyword('else', 'where'),
-      optional($.parenthesized_expression),
+      optional($._parenthesized_expression),
       optional($._block_label),
       $.end_of_statement,
       repeat($._statement)
@@ -1753,7 +1753,7 @@ module.exports = grammar({
       $.derived_type_member_expression,
       $.concatenation_expression,
       $.math_expression,
-      $.parenthesized_expression,
+      $._parenthesized_expression,
       $.call_expression
     )),
 
@@ -1830,14 +1830,14 @@ module.exports = grammar({
       $.concatenation_expression,
       $.math_expression,
       $.unary_expression,
-      $.parenthesized_expression,
+      $._parenthesized_expression,
       $.call_expression,
       $.implied_do_loop_expression,
       $.coarray_expression,
       $.conditional_expression,
     ),
 
-    parenthesized_expression: $ => seq(
+    _parenthesized_expression: $ => seq(
       '(',
       $._expression,
       ')'
diff --git a/test/corpus/constructs.txt b/test/corpus/constructs.txt
index 0e27ee9..7219a8e 100644
--- a/test/corpus/constructs.txt
+++ b/test/corpus/constructs.txt
@@ -863,10 +863,9 @@ end function test
       (intrinsic_type)
       (init_declarator
         (identifier)
-        (parenthesized_expression
-          (relational_expression
-            (identifier)
-            (identifier)))))
+        (relational_expression
+          (identifier)
+          (identifier))))
     (end_of_statement)
     (internal_procedures
       (contains_statement)
diff --git a/test/corpus/expressions.txt b/test/corpus/expressions.txt
index 878be80..eee04ef 100644
--- a/test/corpus/expressions.txt
+++ b/test/corpus/expressions.txt
@@ -610,9 +610,8 @@ END PROGRAM
             (identifier))
           (math_expression
             (identifier)
-            (parenthesized_expression
-              (unary_expression
-                (number_literal)))))
+            (unary_expression
+              (number_literal))))
         (number_literal)))
     (end_of_statement)
     (assignment_statement
@@ -620,10 +619,9 @@ END PROGRAM
       (math_expression
         (number_literal)
         (math_expression
-          (parenthesized_expression
-            (math_expression
-              (identifier)
-              (identifier)))
+          (math_expression
+            (identifier)
+            (identifier))
           (number_literal))))
     (end_of_statement)
     (end_program_statement
@@ -766,25 +764,21 @@ END PROGRAM
     (end_of_statement)
     (assignment_statement
       (identifier)
-      (parenthesized_expression
-        (logical_expression
-          (relational_expression
-            (identifier)
-            (identifier)))))
+      (logical_expression
+        (relational_expression
+          (identifier)
+          (identifier))))
     (end_of_statement)
     (assignment_statement
       (identifier)
       (logical_expression
-        (parenthesized_expression
-          (logical_expression
-            (parenthesized_expression
-              (relational_expression
-                (identifier)
-                (identifier)))
-            (parenthesized_expression
-              (relational_expression
-                (identifier)
-                (identifier)))))
+        (logical_expression
+          (relational_expression
+            (identifier)
+            (identifier))
+          (relational_expression
+            (identifier)
+            (identifier)))
         (boolean_literal)))
     (end_of_statement)
     (assignment_statement
@@ -1121,15 +1115,14 @@ end program array_constructors
       (identifier)
       (array_literal
         (implied_do_loop_expression
-          (parenthesized_expression
-            (math_expression
-              (call_expression
-                (identifier)
-                (argument_list
-                  (math_expression
-                    (identifier)
-                    (number_literal))))
-              (number_literal)))
+          (math_expression
+            (call_expression
+              (identifier)
+              (argument_list
+                (math_expression
+                  (identifier)
+                  (number_literal))))
+            (number_literal))
           (loop_control_expression
             (identifier)
             (number_literal)
@@ -1236,42 +1229,41 @@ end program
       (number_literal))
     (end_of_statement)
     (if_statement
-      (parenthesized_expression
+      (block_if_statement
         (relational_expression
           (identifier)
-          (number_literal)))
-      (end_of_statement)
-      (assignment_statement
-        (identifier)
-        (number_literal))
-      (end_of_statement)
-      (assignment_statement
-        (identifier)
-        (math_expression
+          (number_literal))
+        (end_of_statement)
+        (assignment_statement
           (identifier)
-          (identifier)))
-      (end_of_statement)
-      (elseif_clause
-        (parenthesized_expression
-          (relational_expression
-            (identifier)
-            (number_literal)))
+          (number_literal))
         (end_of_statement)
         (assignment_statement
-          (call_expression
-            (identifier)
-            (argument_list
-              (identifier)))
+          (identifier)
           (math_expression
             (identifier)
             (identifier)))
         (end_of_statement)
-        (print_statement
-          (format_identifier)
-          (output_item_list
-            (identifier)))
-        (end_of_statement))
-      (end_if_statement))
+        (elseif_clause
+          (relational_expression
+            (identifier)
+            (number_literal))
+          (end_of_statement)
+          (assignment_statement
+            (call_expression
+              (identifier)
+              (argument_list
+                (identifier)))
+            (math_expression
+              (identifier)
+              (identifier)))
+          (end_of_statement)
+          (print_statement
+            (format_identifier)
+            (output_item_list
+              (identifier)))
+          (end_of_statement))
+        (end_if_statement)))
     (end_of_statement)
     (end_program_statement
       (end_of_statement))))
diff --git a/test/corpus/regressions.txt b/test/corpus/regressions.txt
index d6b3f37..f445352 100644
--- a/test/corpus/regressions.txt
+++ b/test/corpus/regressions.txt
@@ -97,23 +97,23 @@ end program foo
       (name)
       (end_of_statement))
     (if_statement
-      (block_label_start_expression)
-      (parenthesized_expression
-        (boolean_literal))
-      (end_of_statement)
-      (subroutine_call
-        (identifier)
-        (argument_list))
-      (end_of_statement)
-      (else_clause
-        (block_label)
+      (block_if_statement
+        (block_label_start_expression)
+        (boolean_literal)
         (end_of_statement)
         (subroutine_call
           (identifier)
           (argument_list))
-        (end_of_statement))
-      (end_if_statement
-        (block_label)))
+        (end_of_statement)
+        (else_clause
+          (block_label)
+          (end_of_statement)
+          (subroutine_call
+            (identifier)
+            (argument_list))
+          (end_of_statement))
+        (end_if_statement
+          (block_label))))
     (end_of_statement)
     (end_program_statement
       (name)
@@ -140,14 +140,13 @@ end program
       (name)
       (end_of_statement))
     (if_statement
-      (parenthesized_expression
-        (boolean_literal))
-      (end_of_statement)
-      (elseif_clause
-        (parenthesized_expression
-          (boolean_literal))
-        (end_of_statement))
-      (end_if_statement))
+      (block_if_statement
+        (boolean_literal)
+        (end_of_statement)
+        (elseif_clause
+          (boolean_literal)
+          (end_of_statement))
+        (end_if_statement)))
     (end_of_statement)
     (end_program_statement
       (end_of_statement))))
diff --git a/test/corpus/statements.txt b/test/corpus/statements.txt
index 12c0023..14e2e8a 100644
--- a/test/corpus/statements.txt
+++ b/test/corpus/statements.txt
@@ -860,8 +860,7 @@ END PROGRAM
     (end_of_statement)
     (do_loop_statement
       (while_statement
-        (parenthesized_expression
-          (identifier)))
+        (identifier))
       (end_of_statement)
       (assignment_statement
         (identifier)
@@ -876,10 +875,10 @@ END PROGRAM
         (number_literal))
       (end_of_statement)
       (if_statement
-        (parenthesized_expression
-          (identifier))
-        (keyword_statement
-          (statement_label_reference)))
+        (inline_if_statement
+          (identifier)
+          (keyword_statement
+            (statement_label_reference))))
       (end_of_statement)
       (statement_label)
       (end_do_loop_statement))
@@ -1139,29 +1138,29 @@ END PROGRAM
       (name)
       (end_of_statement))
     (if_statement
-      (parenthesized_expression
+      (inline_if_statement
         (relational_expression
           (identifier)
-          (number_literal)))
-      (assignment_statement
-        (identifier)
-        (number_literal)))
+          (number_literal))
+        (assignment_statement
+          (identifier)
+          (number_literal))))
     (end_of_statement)
     (if_statement
-      (parenthesized_expression
+      (inline_if_statement
         (logical_expression
           (relational_expression
             (identifier)
             (number_literal))
           (relational_expression
             (number_literal)
-            (identifier))))
-      (assignment_statement
-        (identifier)
-        (number_literal)))
+            (identifier)))
+        (assignment_statement
+          (identifier)
+          (number_literal))))
     (end_of_statement)
     (if_statement
-      (parenthesized_expression
+      (block_if_statement
         (relational_expression
           (call_expression
             (identifier)
@@ -1172,14 +1171,13 @@ END PROGRAM
           (call_expression
             (identifier)
             (argument_list
-              (string_literal)))))
-      (end_of_statement)
-      (assignment_statement
-        (identifier)
-        (number_literal))
-      (end_of_statement)
-      (elseif_clause
-        (parenthesized_expression
+              (string_literal))))
+        (end_of_statement)
+        (assignment_statement
+          (identifier)
+          (number_literal))
+        (end_of_statement)
+        (elseif_clause
           (relational_expression
             (call_expression
               (identifier)
@@ -1193,14 +1191,13 @@ END PROGRAM
                 (call_expression
                   (identifier)
                   (argument_list
-                    (number_literal)))))))
-        (end_of_statement)
-        (assignment_statement
-          (identifier)
-          (number_literal))
-        (end_of_statement))
-      (elseif_clause
-        (parenthesized_expression
+                    (number_literal))))))
+          (end_of_statement)
+          (assignment_statement
+            (identifier)
+            (number_literal))
+          (end_of_statement))
+        (elseif_clause
           (relational_expression
             (call_expression
               (identifier)
@@ -1214,68 +1211,67 @@ END PROGRAM
                 (call_expression
                   (identifier)
                   (argument_list
-                    (number_literal)))))))
-        (end_of_statement))
-      (else_clause
-        (end_of_statement)
-        (assignment_statement
-          (identifier)
-          (number_literal))
-        (end_of_statement))
-      (end_if_statement))
+                    (number_literal))))))
+          (end_of_statement))
+        (else_clause
+          (end_of_statement)
+          (assignment_statement
+            (identifier)
+            (number_literal))
+          (end_of_statement))
+        (end_if_statement)))
     (end_of_statement)
     (if_statement
-      (block_label_start_expression)
-      (parenthesized_expression
+      (block_if_statement
+        (block_label_start_expression)
         (relational_expression
           (identifier)
-          (number_literal)))
-      (end_of_statement)
-      (assignment_statement
-        (identifier)
-        (number_literal))
-      (end_of_statement)
-      (elseif_clause
-        (parenthesized_expression
-          (relational_expression
-            (identifier)
-            (number_literal)))
-        (block_label)
+          (number_literal))
         (end_of_statement)
         (assignment_statement
           (identifier)
           (number_literal))
         (end_of_statement)
-        (if_statement
-          (parenthesized_expression
-            (call_expression
-              (identifier)
-              (argument_list
-                (extent_specifier
-                  (number_literal)
-                  (number_literal)))))
+        (elseif_clause
+          (relational_expression
+            (identifier)
+            (number_literal))
+          (block_label)
           (end_of_statement)
           (assignment_statement
             (identifier)
             (number_literal))
           (end_of_statement)
-          (else_clause
-            (end_of_statement)
-            (assignment_statement
-              (identifier)
-              (number_literal))
-            (end_of_statement))
-          (end_if_statement))
-        (end_of_statement))
-      (else_clause
-        (block_label)
-        (end_of_statement)
-        (assignment_statement
-          (identifier)
-          (number_literal))
-        (end_of_statement))
-      (end_if_statement
-        (block_label)))
+          (if_statement
+            (block_if_statement
+              (call_expression
+                (identifier)
+                (argument_list
+                  (extent_specifier
+                    (number_literal)
+                    (number_literal))))
+              (end_of_statement)
+              (assignment_statement
+                (identifier)
+                (number_literal))
+              (end_of_statement)
+              (else_clause
+                (end_of_statement)
+                (assignment_statement
+                  (identifier)
+                  (number_literal))
+                (end_of_statement))
+              (end_if_statement)))
+          (end_of_statement))
+        (else_clause
+          (block_label)
+          (end_of_statement)
+          (assignment_statement
+            (identifier)
+            (number_literal))
+          (end_of_statement))
+        (end_if_statement
+          (block_label))))
     (end_of_statement)
     (end_program_statement
       (end_of_statement))))
@@ -1315,37 +1311,25 @@ end program
       (name)
       (end_of_statement))
     (where_statement
-      (parenthesized_expression
+      (inline_where_statement
         (relational_expression
           (identifier)
-          (number_literal)))
-      (assignment_statement
-        (identifier)
-        (math_expression
+          (number_literal))
+        (assignment_statement
           (identifier)
-          (identifier))))
+          (math_expression
+            (identifier)
+            (identifier)))))
     (end_of_statement)
     (where_statement
-      (parenthesized_expression
+      (block_where_statement
         (relational_expression
           (identifier)
-          (number_literal)))
-      (end_of_statement)
-      (assignment_statement
-        (identifier)
-        (math_expression
-          (identifier)
-          (number_literal)))
-      (end_of_statement)
-      (assignment_statement
-        (identifier)
-        (math_expression
+          (number_literal))
+        (end_of_statement)
+        (assignment_statement
           (identifier)
-          (number_literal)))
-      (end_of_statement)
-      (elsewhere_clause
-        (parenthesized_expression
-          (relational_expression
+          (math_expression
             (identifier)
             (number_literal)))
         (end_of_statement)
@@ -1354,50 +1338,60 @@ end program
           (math_expression
             (identifier)
             (number_literal)))
-        (end_of_statement))
-      (elsewhere_clause
         (end_of_statement)
-        (assignment_statement
-          (identifier)
-          (boolean_literal))
-        (end_of_statement))
-      (end_where_statement))
+        (elsewhere_clause
+          (relational_expression
+            (identifier)
+            (number_literal))
+          (end_of_statement)
+          (assignment_statement
+            (identifier)
+            (math_expression
+              (identifier)
+              (number_literal)))
+          (end_of_statement))
+        (elsewhere_clause
+          (end_of_statement)
+          (assignment_statement
+            (identifier)
+            (boolean_literal))
+          (end_of_statement))
+        (end_where_statement)))
     (end_of_statement)
     (where_statement
-      (block_label_start_expression)
-      (parenthesized_expression
+      (block_where_statement
+        (block_label_start_expression)
         (relational_expression
           (identifier)
-          (number_literal)))
-      (end_of_statement)
-      (assignment_statement
-        (identifier)
-        (math_expression
-          (identifier)
-          (number_literal)))
-      (end_of_statement)
-      (elsewhere_clause
-        (parenthesized_expression
-          (relational_expression
-            (identifier)
-            (number_literal)))
-        (block_label)
+          (number_literal))
         (end_of_statement)
         (assignment_statement
           (identifier)
           (math_expression
             (identifier)
             (number_literal)))
-        (end_of_statement))
-      (elsewhere_clause
-        (block_label)
         (end_of_statement)
-        (assignment_statement
-          (identifier)
-          (boolean_literal))
-        (end_of_statement))
-      (end_where_statement
-        (block_label)))
+        (elsewhere_clause
+          (relational_expression
+            (identifier)
+            (number_literal))
+          (block_label)
+          (end_of_statement)
+          (assignment_statement
+            (identifier)
+            (math_expression
+              (identifier)
+              (number_literal)))
+          (end_of_statement))
+        (elsewhere_clause
+          (block_label)
+          (end_of_statement)
+          (assignment_statement
+            (identifier)
+            (boolean_literal))
+          (end_of_statement))
+        (end_where_statement
+          (block_label))))
     (end_of_statement)
     (end_program_statement
       (end_of_statement))))
@@ -1432,167 +1426,171 @@ END PROGRAM
       (name)
       (end_of_statement))
     (forall_statement
-      (triplet_spec
-        (identifier)
-        (number_literal)
-        (identifier))
-      (triplet_spec
-        (identifier)
-        (number_literal)
-        (identifier))
-      (relational_expression
-        (call_expression
+      (inline_forall_statement
+        (triplet_spec
           (identifier)
-          (argument_list
-            (identifier)
-            (identifier)))
-        (number_literal))
-      (assignment_statement
-        (call_expression
+          (number_literal)
+          (identifier))
+        (triplet_spec
           (identifier)
-          (argument_list
-            (identifier)
-            (identifier)))
-        (math_expression
           (number_literal)
+          (identifier))
+        (relational_expression
           (call_expression
             (identifier)
             (argument_list
               (identifier)
-              (identifier))))))
-    (end_of_statement)
-    (forall_statement
-      (triplet_spec
-        (identifier)
-        (number_literal)
-        (number_literal))
-      (pointer_association_statement
-        (derived_type_member_expression
+              (identifier)))
+          (number_literal))
+        (assignment_statement
           (call_expression
             (identifier)
             (argument_list
+              (identifier)
               (identifier)))
-          (type_member))
-        (call_expression
-          (identifier)
-          (argument_list
-            (math_expression
-              (number_literal)
-              (call_expression
+          (math_expression
+            (number_literal)
+            (call_expression
+              (identifier)
+              (argument_list
                 (identifier)
-                (argument_list
-                  (math_expression
-                    (identifier)
-                    (number_literal))
-                  (number_literal))))))))
+                (identifier)))))))
     (end_of_statement)
     (forall_statement
-      (triplet_spec
-        (identifier)
-        (number_literal)
-        (identifier))
-      (triplet_spec
-        (identifier)
-        (number_literal)
-        (identifier))
-      (end_of_statement)
-      (where_statement
-        (parenthesized_expression
-          (relational_expression
+      (inline_forall_statement
+        (triplet_spec
+          (identifier)
+          (number_literal)
+          (number_literal))
+        (pointer_association_statement
+          (derived_type_member_expression
             (call_expression
               (identifier)
               (argument_list
-                (identifier)
                 (identifier)))
-            (number_literal)))
-        (assignment_statement
+            (type_member))
           (call_expression
             (identifier)
             (argument_list
-              (identifier)
-              (identifier)))
-          (math_expression
-            (number_literal)
-            (call_expression
-              (identifier)
-              (argument_list
-                (identifier)
-                (identifier))))))
-      (end_of_statement)
-      (end_forall_statement))
+              (math_expression
+                (number_literal)
+                (call_expression
+                  (identifier)
+                  (argument_list
+                    (math_expression
+                      (identifier)
+                      (number_literal))
+                    (number_literal)))))))))
     (end_of_statement)
     (forall_statement
-      (triplet_spec
-        (identifier)
-        (number_literal)
-        (math_expression
+      (block_forall_statement
+        (triplet_spec
           (identifier)
-          (number_literal)))
-      (triplet_spec
-        (identifier)
-        (number_literal)
-        (math_expression
+          (number_literal)
+          (identifier))
+        (triplet_spec
           (identifier)
-          (number_literal)))
-      (relational_expression
-        (call_expression
+          (number_literal)
+          (identifier))
+        (end_of_statement)
+        (where_statement
+          (inline_where_statement
+            (relational_expression
+              (call_expression
+                (identifier)
+                (argument_list
+                  (identifier)
+                  (identifier)))
+              (number_literal))
+            (assignment_statement
+              (call_expression
+                (identifier)
+                (argument_list
+                  (identifier)
+                  (identifier)))
+              (math_expression
+                (number_literal)
+                (call_expression
+                  (identifier)
+                  (argument_list
+                    (identifier)
+                    (identifier)))))))
+        (end_of_statement)
+        (end_forall_statement)))
+    (end_of_statement)
+    (forall_statement
+      (block_forall_statement
+        (triplet_spec
           (identifier)
-          (argument_list
+          (number_literal)
+          (math_expression
             (identifier)
-            (identifier)))
-        (number_literal))
-      (end_of_statement)
-      (assignment_statement
-        (call_expression
+            (number_literal)))
+        (triplet_spec
           (identifier)
-          (argument_list
+          (number_literal)
+          (math_expression
             (identifier)
-            (identifier)))
-        (math_expression
+            (number_literal)))
+        (relational_expression
+          (call_expression
+            (identifier)
+            (argument_list
+              (identifier)
+              (identifier)))
+          (number_literal))
+        (end_of_statement)
+        (assignment_statement
+          (call_expression
+            (identifier)
+            (argument_list
+              (identifier)
+              (identifier)))
           (math_expression
             (math_expression
-              (call_expression
-                (identifier)
-                (argument_list
+              (math_expression
+                (call_expression
                   (identifier)
-                  (math_expression
+                  (argument_list
                     (identifier)
-                    (number_literal))))
+                    (math_expression
+                      (identifier)
+                      (number_literal))))
+                (call_expression
+                  (identifier)
+                  (argument_list
+                    (identifier)
+                    (math_expression
+                      (identifier)
+                      (number_literal)))))
               (call_expression
                 (identifier)
                 (argument_list
-                  (identifier)
                   (math_expression
                     (identifier)
-                    (number_literal)))))
+                    (number_literal))
+                  (identifier))))
             (call_expression
               (identifier)
               (argument_list
                 (math_expression
                   (identifier)
                   (number_literal))
-                (identifier))))
+                (identifier)))))
+        (end_of_statement)
+        (assignment_statement
           (call_expression
             (identifier)
             (argument_list
-              (math_expression
-                (identifier)
-                (number_literal))
-              (identifier)))))
-      (end_of_statement)
-      (assignment_statement
-        (call_expression
-          (identifier)
-          (argument_list
-            (identifier)
-            (identifier)))
-        (call_expression
-          (identifier)
-          (argument_list
+              (identifier)
+              (identifier)))
+          (call_expression
             (identifier)
-            (identifier))))
-      (end_of_statement)
-      (end_forall_statement))
+            (argument_list
+              (identifier)
+              (identifier))))
+        (end_of_statement)
+        (end_forall_statement)))
     (end_of_statement)
     (end_program_statement
       (end_of_statement))))
@@ -2962,16 +2960,14 @@ end program
     (end_of_statement)
     (file_position_statement
       (unit_identifier
-        (parenthesized_expression
-          (number_literal))))
+        (number_literal)))
     (end_of_statement)
     (file_position_statement
       (unit_identifier
-        (parenthesized_expression
-          (call_expression
-            (identifier)
-            (argument_list
-              (identifier))))))
+        (call_expression
+          (identifier)
+          (argument_list
+            (identifier)))))
     (end_of_statement)
     (file_position_statement
       (string_literal))
@@ -3119,8 +3115,7 @@ end
 (translation_unit
   (program
     (arithmetic_if_statement
-      (parenthesized_expression
-        (identifier))
+      (identifier)
       (statement_label_reference)
       (statement_label_reference)
       (statement_label_reference))
@@ -3390,19 +3385,19 @@ end program test
         (identifier)))
     (end_of_statement)
     (if_statement
-      (parenthesized_expression
+      (inline_if_statement
         (relational_expression
           (identifier)
-          (number_literal)))
-      (assignment_statement
-        (coarray_expression
-          (identifier)
-          (coarray_index
-            (number_literal)
-            (keyword_argument
-              (identifier)
-              (identifier))))
-        (identifier)))
+          (number_literal))
+        (assignment_statement
+          (coarray_expression
+            (identifier)
+            (coarray_index
+              (number_literal)
+              (keyword_argument
+                (identifier)
+                (identifier))))
+          (identifier))))
     (end_of_statement)
     (allocate_statement
       (coarray_allocation
